clippy_lints/src/literal_string_with_formatting_arg.rs

```diff
@@ -0,0 +1,93 @@
+use rustc_ast::ast::{Expr, ExprKind};
+use rustc_ast::token::LitKind;
+use rustc_lint::{EarlyContext, EarlyLintPass};
+use rustc_session::impl_lint_pass;
+use rustc_span::BytePos;
+
+use clippy_utils::diagnostics::span_lint;
+use regex::Regex;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks if string literals have formatting arguments outside of macros
+    /// using them (like `format!`).
+    ///
+    /// ### Why is this bad?
+    /// It will likely not generate the expected content.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let x: Option<usize> = None;
+    /// let y = "hello";
+    /// x.expect("{y:?}");
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let x: Option<usize> = None;
+    /// let y = "hello";
+    /// x.expect(&format!("{y:?}"));
+    /// ```
+    #[clippy::version = "1.83.0"]
+    pub LITERAL_STRING_WITH_FORMATTING_ARG,
+    suspicious,
+    "Checks if string literals have formatting arguments"
+}
+
+pub struct LiteralStringWithFormattingArg {
+    regex: Regex,
+}
+
+impl LiteralStringWithFormattingArg {
+    pub fn new() -> Self {
+        Self {
+            regex: Regex::new(r"([^\{]\{\w+:?[^\}]*\})|([^\{]\{})|(^\{\w+:?[^\}]*\})|(^\{})").unwrap(),
+        }
+    }
+}
+
+impl_lint_pass!(LiteralStringWithFormattingArg => [LITERAL_STRING_WITH_FORMATTING_ARG]);
+
+impl EarlyLintPass for LiteralStringWithFormattingArg {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &Expr) {
+        if let ExprKind::Lit(lit) = expr.kind {
+            let add = match lit.kind {
+                LitKind::Str => BytePos(1),
+                LitKind::StrRaw(nb) => BytePos((nb + 1).into()),
+                _ => return,
+            };
+            let lit_s = lit.symbol.as_str();
+            let lo = expr.span.lo();
+            let spans = self
+                .regex
+                .find_iter(lit_s)
+                .map(|match_| {
+                    let mut span = expr
+                        .span
+                        .with_hi(lo + BytePos(match_.end() as _) + add)
+                        .with_lo(lo + BytePos(match_.start() as _) + add);
+                    // The regex captures the character before the `{` if it's not a newline character,
+                    // so in this case, we need to skip it.
+                    if !match_.as_str().starts_with('{') {
+                        span = span.with_lo(span.lo() + BytePos(1));
+                    }
+                    span
+                })
+                .collect::<Vec<_>>();
+            if spans.len() == 1 {
+                span_lint(
```

Comment:
I could emit a suggestion for both cases by wrapping the literal into `&format!()`. Not sure if it's a good idea since it could be wrong in a lot of cases so for now I didn't do it.